### PR TITLE
Remove producer-is-in-world assumptions from Production

### PR DESCRIPTION
**Pull request by penev92**
Justification:

> D2k upgrades need to be actors that never actor the world, produced by the PlayerActor, who is also not in the world, but the Production trait expects the producer to be in the world even if it has `Exit.MoveIntoWorld = False`, allowing the producees to not enter the world.

Use-case resides on @OmegaBolt's [d2k balancing branch](https://github.com/OmegaBolt/OpenRA).
Anything currently in bleed should work exactly the same as it does now.

### CODE REVIEW

**Comment by OmegaBolt** [comment]
Tested this on the my branch and the upgrades now finish production and 'exit' without a crash, as expected.

**Comment by Mailaender** [comment]
Current d2k updates and spy induced veterancy still works. :white_check_mark: Also no regressions on the current exit point behavior.

File: OpenRA.Mods.Common/Traits/Production.cs [hunk 1]
```diff
@@ -35,39 +35,50 @@ public class Production
 		public ProductionInfo Info;
 		public string Faction { get; private set; }
 
+		readonly bool occupiesSpace;
+
 		public Production(ActorInitializer init, ProductionInfo info)
 		{
 			Info = info;
+			occupiesSpace = init.Self.Info.Traits.WithInterface<IOccupySpaceInfo>().Any();
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Production.cs]: This could have been simplified to `.Contains<>`.

File: OpenRA.Mods.Common/Traits/Production.cs [hunk 2]
```diff
@@ -107,7 +118,13 @@ public virtual bool Produce(Actor self, ActorInfo producee, string factionVarian
 			if (Reservable.IsReserved(self))
 				return false;
 
-			// pick a spawn/exit point pair
+			if (!occupiesSpace)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Production.cs]: This unnecessarily duplicates the block below.

**Pull request merged**